scripts/suite.py

```diff
@@ -71,6 +71,7 @@
                               <suite_branch> to be ignored for scheduling
                               purposes, but it will still be used for test
                               running.
+  --no-validate-refs          Don't validate that git SHA1s passed to -S exist.
```

Comment:
Could you change this to `--validate-refs` with `default: True`? Then you would avoid the weird "double-negative" later on.

Reply:
Could you please rename `--no-validate-refs` to `--validate-sha1` defaults: 0, because:
1) Options should not have negative notation.
2) This does not produces double-negative coding.
3) sha1 is better since this option is directly related to SHA1 options.
It is better to put the description for this key next to:
```-S <sha1>, --sha1 <sha1>    The ceph sha1 to run against (overrides -c)```.

Reply:
I've tried changing this to `--validate-refs` with the default as `True`, but then i can't figure out how to make it False at runtime. My option definition looks like this:

```
  --validate-refs             Validate that git SHA1s passed to -S exist.
                              [default: True]
```

but I've tried a few ways to set it and none of them work:

```console
$ teuthology-suite -c luminous -S abcd1234 --suite rbd:basic -d ubuntu -D 14.04 -m ubuntu --suite-branch luminous --validate-refs=False /home/teuthology/overrides.yml
--validate-refs must not have an argument
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
$ teuthology-suite -c luminous -S abcd1234 --suite rbd:basic -d ubuntu -D 14.04 -m ubuntu --suite-branch luminous --no-validate-refs /home/teuthology/overrides.yml
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
$ teuthology-suite -c luminous -S abcd1234 --suite rbd:basic -d ubuntu -D 14.04 -m ubuntu --suite-branch luminous --without-validate-refs /home/teuthology/overrides.yml
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
```

Any thoughts on how this is supposed to work?

Reply:
If you look carefully you can see that '=' is not used anywhere with arguments.

Reply:
@adamwg I still would like to ask you to rename argument to `--validate_sha1` to avoid future confusion and recall easily the arguments.

Reply:
@kshtsk Not sure you're understanding the problem I'm having. If I remove the `=` I still have the same issue. Updated the code to this:

```
  --validate-sha1 <validate_sha1>
                              Validate that git SHA1s passed to -S exist.
                              [default: 0]
```

but neither of the following makes the option false:

```console
$ teuthology-suite -c luminous -S abcd1234 --suite rbd:basic -d ubuntu -D 14.04 -m ubuntu --suite-branch luminous --validate-sha1 False /home/teuthology/overrides.yml
--validate-refs must not have an argument
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
$ teuthology-suite -c luminous -S abcd1234 --suite rbd:basic -d ubuntu -D 14.04 -m ubuntu --suite-branch luminous --validate-sha1 0 /home/teuthology/overrides.yml
usage: teuthology-suite --help
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --suite <suite> [options] [<config_yaml>...]
       teuthology-suite [-v | -vv ] --machine-type <type> --ceph <ceph> --rerun <name>  [options] [<config_yaml>...]
```

It seems like a bug in this silly option parsing library that it can't create a bool flag with true as a default. Unless someone has a solution, I think the only way to go is with my original solution, naming the flag `no-validate-sha1`.

Reply:
@adamwg It is hard to say what the bug you have not seeing the changes. Seems like you changed option to `--validate-sha1` but still trying to use somewhere `--validate-refs`.
See example, it works fine:
```
import docopt
import sys

doc = """
usage: t.py --help
       t.py [options]

Test arguments.

arguments:
  -b <bln>, --bool <bln> 
                                Boolean value [default: True]
"""

def main(argv=sys.argv[1:]):
    args = docopt.docopt(doc, argv=argv)
    print("Arguments: %s" % args)

if __name__ == "__main__":
    main()
```
1) Default value should is True for ```python t.py```:
```
Arguments: {'--bool': 'True',
 '--help': False}
```
2) Value is changed to False for ```python t.py -b False```
```
Arguments: {'--bool': 'False',
 '--help': False}
```
And, maybe, your problem is that you do not provide **mandatory** arguments:  `--machine-type`, `--ceph` and `--suite` (or `--rerun`).

Reply:
> I've tried changing this to `--validate-refs` with the default as `True`, but then i can't figure out how to make it False at runtime. My option definition looks like this:
> 
> ```
>   --validate-refs             Validate that git SHA1s passed to -S exist.
>                               [default: True]
> ```
> but I've tried a few ways to set it and none of them work:
> .
> .
> .
> --validate-refs must not have an argument
Looks like you have left this `--validate-refs` declaration somewhere, which is syntactically  incorrect, because you didn't provide a placeholder for value in angle brackets but declaring the default value.

Reply:
@kshtsk In your example, the `--bool` argument value is not a bool, it's a string. If you do something like `if bool` you'll find it's always truey (unless you set it to an empty string, but that's a totally unintuitive way to make a boolean option false).

That did help me figure out the issue, though - it's always treating the argument as a string, so we need to convert it manually. See the latest update to this PR for a working solution.

Reply:
very true, if the default value is not repeatable it is interpreted as a string